src/subject/messages.rs

```diff
@@ -1,54 +1,88 @@
+use crate::database::Executor;
 use actix_web::HttpResponse;
 use async_trait::async_trait;
 use serde::{Deserialize, Serialize};
 
-use super::model::{Subjects, SubjectsError};
 use crate::database::Connection;
 use crate::message::MessageResponder;
 
 #[derive(Deserialize, Serialize)]
 #[serde(tag = "type", content = "payload")]
 pub enum SubjectsMessage {
-    SubjectsQuery(SubjectsQuery),
+    SubjectsQuery(Option<serde_json::Value>),
 }
 
 #[async_trait]
 impl MessageResponder for SubjectsMessage {
     #[allow(clippy::async_yields_async)]
     async fn handle(&self, connection: Connection<'_, '_>) -> HttpResponse {
         match self {
-            SubjectsMessage::SubjectsQuery(message) => message.handle(connection).await,
+            SubjectsMessage::SubjectsQuery(_) => subjects_query(connection).await,
         }
     }
 }
 
-#[derive(Deserialize, Serialize)]
-#[serde(rename_all = "camelCase")]
-pub struct SubjectsQuery {}
-
-#[async_trait]
-impl MessageResponder for SubjectsQuery {
-    #[allow(clippy::async_yields_async)]
-    async fn handle(&self, connection: Connection<'_, '_>) -> HttpResponse {
-        let subjects = match connection {
-            Connection::Pool(pool) => Subjects::fetch(pool).await,
-            Connection::Transaction(transaction) => {
-                Subjects::fetch_via_transaction(transaction).await
-            }
-        };
-        match subjects {
-            Ok(data) => HttpResponse::Ok()
-                .content_type("application/json; charset=utf-8")
-                .json(&data),
-            Err(e) => {
-                println!("/subjects: {:?}", e);
-                match e {
-                    SubjectsError::DatabaseError { .. } => {
-                        HttpResponse::InternalServerError().finish()
-                    }
-                    SubjectsError::InvalidInstance => HttpResponse::InternalServerError().finish(),
-                }
-            }
+async fn subjects_query(connection: Connection<'_, '_>) -> HttpResponse {
+    let subjects = match connection {
+        Connection::Pool(pool) => fetch_subjects(pool).await,
+        Connection::Transaction(transaction) => fetch_subjects(transaction).await,
+    };
+    match subjects {
+        Ok(data) => HttpResponse::Ok()
+            .content_type("application/json; charset=utf-8")
+            .json(&data),
+        Err(e) => {
+            println!("/subjects: {:?}", e);
+            HttpResponse::InternalServerError().finish()
         }
     }
 }
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Subjects {
+    pub subjects: Vec<Subject>,
+}
```

Comment:
For consistency, I'd keep that in `model`. This way, we have a clear separation of concerns:
- `messages` defines the accepted messages and how they handle possible errors.
- `model` does the internal stuff, i.e. talk to database etc.

Reply:
I think we need to talk about this in a meeting. The split can be improved. Before the refactoring the struct for the input and for the output are in different files which does not seem to be the best solution. The quoted struct for example is so specific to the query that it cannot be used anywhere else...